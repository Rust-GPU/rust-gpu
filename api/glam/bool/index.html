<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`bool` vector mask types. "><title>glam::bool - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="glam" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-e32f0c247825364d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../glam/index.html">glam</a><span class="version">0.29.0</span></h2></div><h2 class="location"><a href="#">Module bool</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#functions">Functions</a></li></ul></section><h2><a href="../index.html">In crate glam</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../glam/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">glam</a>::<wbr><a class="mod" href="#">bool</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../src/glam/bool.rs.html#1-126">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p><code>bool</code> vector mask types. </p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.BVec2.html" title="struct glam::bool::BVec2">BVec2</a></div><div class="desc docblock-short">A 2-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="struct" href="struct.BVec3.html" title="struct glam::bool::BVec3">BVec3</a></div><div class="desc docblock-short">A 3-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="struct" href="struct.BVec3A.html" title="struct glam::bool::BVec3A">BVec3A</a></div><div class="desc docblock-short">A 3-dimensional SIMD vector mask.</div></li><li><div class="item-name"><a class="struct" href="struct.BVec4.html" title="struct glam::bool::BVec4">BVec4</a></div><div class="desc docblock-short">A 4-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="struct" href="struct.BVec4A.html" title="struct glam::bool::BVec4A">BVec4A</a></div><div class="desc docblock-short">A 4-dimensional SIMD vector mask.</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.bvec2.html" title="fn glam::bool::bvec2">bvec2</a></div><div class="desc docblock-short">Creates a 2-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="fn" href="fn.bvec3.html" title="fn glam::bool::bvec3">bvec3</a></div><div class="desc docblock-short">Creates a 3-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="fn" href="fn.bvec3a.html" title="fn glam::bool::bvec3a">bvec3a</a></div><div class="desc docblock-short">Creates a 3-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="fn" href="fn.bvec4.html" title="fn glam::bool::bvec4">bvec4</a></div><div class="desc docblock-short">Creates a 4-dimensional <code>bool</code> vector mask.</div></li><li><div class="item-name"><a class="fn" href="fn.bvec4a.html" title="fn glam::bool::bvec4a">bvec4a</a></div><div class="desc docblock-short">Creates a 4-dimensional <code>bool</code> vector mask.</div></li></ul></section></div></main></body></html>