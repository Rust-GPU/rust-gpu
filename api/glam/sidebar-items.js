window.SIDEBAR_ITEMS = {"enum":["EulerRot"],"mod":["bool","f32","f64","i16","i32","i64","i8","swizzles","u16","u32","u64","u8"],"trait":["FloatExt"]};