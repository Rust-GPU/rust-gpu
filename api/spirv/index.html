<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The SPIR-V header."><title>spirv - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv/index.html">spirv</a><span class="version">0.3.0+sdk-1.3.268.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">spirv</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/spirv/lib.rs.html#1-14">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The SPIR-V header.</p>
<p>This crate contains Rust definitions of all SPIR-V structs, enums,
and constants.</p>
<p>The version of this crate is the version of SPIR-V it contains.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.CooperativeMatrixOperands.html" title="struct spirv::CooperativeMatrixOperands">CooperativeMatrixOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_operands_a_cooperative_matrix_operands">CooperativeMatrixOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.FPFastMathMode.html" title="struct spirv::FPFastMathMode">FPFastMathMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode">FPFastMathMode</a></div></li><li><div class="item-name"><a class="struct" href="struct.FragmentShadingRate.html" title="struct spirv::FragmentShadingRate">FragmentShadingRate</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fragment_shading_rate_a_fragment_shading_rate">FragmentShadingRate</a></div></li><li><div class="item-name"><a class="struct" href="struct.FunctionControl.html" title="struct spirv::FunctionControl">FunctionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_control_a_function_control">FunctionControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.ImageOperands.html" title="struct spirv::ImageOperands">ImageOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_operands_a_image_operands">ImageOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.KernelProfilingInfo.html" title="struct spirv::KernelProfilingInfo">KernelProfilingInfo</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_profiling_info_a_kernel_profiling_info">KernelProfilingInfo</a></div></li><li><div class="item-name"><a class="struct" href="struct.LoopControl.html" title="struct spirv::LoopControl">LoopControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_loop_control_a_loop_control">LoopControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemoryAccess.html" title="struct spirv::MemoryAccess">MemoryAccess</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_access_a_memory_access">MemoryAccess</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemorySemantics.html" title="struct spirv::MemorySemantics">MemorySemantics</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics_a_memory_semantics">MemorySemantics</a></div></li><li><div class="item-name"><a class="struct" href="struct.RayFlags.html" title="struct spirv::RayFlags">RayFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_flags_a_ray_flags">RayFlags</a></div></li><li><div class="item-name"><a class="struct" href="struct.SelectionControl.html" title="struct spirv::SelectionControl">SelectionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_selection_control_a_selection_control">SelectionControl</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.AccessQualifier.html" title="enum spirv::AccessQualifier">AccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_access_qualifier_a_access_qualifier">AccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.AddressingModel.html" title="enum spirv::AddressingModel">AddressingModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_addressing_model_a_addressing_model">AddressingModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.BuiltIn.html" title="enum spirv::BuiltIn">BuiltIn</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_built_in_a_built_in">BuiltIn</a></div></li><li><div class="item-name"><a class="enum" href="enum.CLOp.html" title="enum spirv::CLOp">CLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.ExtendedInstructionSet.100.html">OpenCL.std</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.Capability.html" title="enum spirv::Capability">Capability</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_capability_a_capability">Capability</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixLayout.html" title="enum spirv::CooperativeMatrixLayout">CooperativeMatrixLayout</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_layout_a_cooperative_matrix_layout">CooperativeMatrixLayout</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixUse.html" title="enum spirv::CooperativeMatrixUse">CooperativeMatrixUse</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_use_a_cooperative_matrix_use">CooperativeMatrixUse</a></div></li><li><div class="item-name"><a class="enum" href="enum.Decoration.html" title="enum spirv::Decoration">Decoration</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_decoration_a_decoration">Decoration</a></div></li><li><div class="item-name"><a class="enum" href="enum.Dim.html" title="enum spirv::Dim">Dim</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_dim_a_dim">Dim</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionMode.html" title="enum spirv::ExecutionMode">ExecutionMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_mode_a_execution_mode">ExecutionMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionModel.html" title="enum spirv::ExecutionModel">ExecutionModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_model_a_execution_model">ExecutionModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPDenormMode.html" title="enum spirv::FPDenormMode">FPDenormMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_denorm_mode_a_fp_denorm_mode">FPDenormMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPOperationMode.html" title="enum spirv::FPOperationMode">FPOperationMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_operation_mode_a_fp_operation_mode">FPOperationMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPRoundingMode.html" title="enum spirv::FPRoundingMode">FPRoundingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_rounding_mode_a_fp_rounding_mode">FPRoundingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FunctionParameterAttribute.html" title="enum spirv::FunctionParameterAttribute">FunctionParameterAttribute</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_parameter_attribute_a_function_parameter_attribute">FunctionParameterAttribute</a></div></li><li><div class="item-name"><a class="enum" href="enum.GLOp.html" title="enum spirv::GLOp">GLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/GLSL.std.450.html">GLSL.std.450</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.GroupOperation.html" title="enum spirv::GroupOperation">GroupOperation</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_group_operation_a_group_operation">GroupOperation</a></div></li><li><div class="item-name"><a class="enum" href="enum.HostAccessQualifier.html" title="enum spirv::HostAccessQualifier">HostAccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_host_access_qualifier_a_host_access_qualifier">HostAccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelDataType.html" title="enum spirv::ImageChannelDataType">ImageChannelDataType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_data_type_a_image_channel_data_type">ImageChannelDataType</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelOrder.html" title="enum spirv::ImageChannelOrder">ImageChannelOrder</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_order_a_image_channel_order">ImageChannelOrder</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageFormat.html" title="enum spirv::ImageFormat">ImageFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_format_a_image_format">ImageFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.InitializationModeQualifier.html" title="enum spirv::InitializationModeQualifier">InitializationModeQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_initialization_mode_qualifier_a_initialization_mode_qualifier">InitializationModeQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.KernelEnqueueFlags.html" title="enum spirv::KernelEnqueueFlags">KernelEnqueueFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_enqueue_flags_a_kernel_enqueue_flags">KernelEnqueueFlags</a></div></li><li><div class="item-name"><a class="enum" href="enum.LinkageType.html" title="enum spirv::LinkageType">LinkageType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_linkage_type_a_linkage_type">LinkageType</a></div></li><li><div class="item-name"><a class="enum" href="enum.LoadCacheControl.html" title="enum spirv::LoadCacheControl">LoadCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_load_cache_control_a_load_cache_control">LoadCacheControl</a></div></li><li><div class="item-name"><a class="enum" href="enum.MemoryModel.html" title="enum spirv::MemoryModel">MemoryModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_model_a_memory_model">MemoryModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.Op.html" title="enum spirv::Op">Op</a></div><div class="desc docblock-short">SPIR-V <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_instructions_a_instructions">instructions</a> opcodes</div></li><li><div class="item-name"><a class="enum" href="enum.OverflowModes.html" title="enum spirv::OverflowModes">OverflowModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_overflow_modes_a_overflow_modes">OverflowModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.PackedVectorFormat.html" title="enum spirv::PackedVectorFormat">PackedVectorFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_packed_vector_format_a_packed_vector_format">PackedVectorFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.QuantizationModes.html" title="enum spirv::QuantizationModes">QuantizationModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_quantization_modes_a_quantization_modes">QuantizationModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCandidateIntersectionType.html" title="enum spirv::RayQueryCandidateIntersectionType">RayQueryCandidateIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_candidate_intersection_type_a_ray_query_candidate_intersection_type">RayQueryCandidateIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCommittedIntersectionType.html" title="enum spirv::RayQueryCommittedIntersectionType">RayQueryCommittedIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_committed_intersection_type_a_ray_query_committed_intersection_type">RayQueryCommittedIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryIntersection.html" title="enum spirv::RayQueryIntersection">RayQueryIntersection</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_intersection_a_ray_query_intersection">RayQueryIntersection</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerAddressingMode.html" title="enum spirv::SamplerAddressingMode">SamplerAddressingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_addressing_mode_a_sampler_addressing_mode">SamplerAddressingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerFilterMode.html" title="enum spirv::SamplerFilterMode">SamplerFilterMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_filter_mode_a_sampler_filter_mode">SamplerFilterMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.Scope.html" title="enum spirv::Scope">Scope</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_scope_a_scope">Scope</a></div></li><li><div class="item-name"><a class="enum" href="enum.SourceLanguage.html" title="enum spirv::SourceLanguage">SourceLanguage</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_source_language_a_source_language">SourceLanguage</a></div></li><li><div class="item-name"><a class="enum" href="enum.StorageClass.html" title="enum spirv::StorageClass">StorageClass</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_storage_class_a_storage_class">StorageClass</a></div></li><li><div class="item-name"><a class="enum" href="enum.StoreCacheControl.html" title="enum spirv::StoreCacheControl">StoreCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_store_cache_control_a_store_cache_control">StoreCacheControl</a></div></li></ul><h2 id="constants" class="section-header">Constants<a href="#constants" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="constant" href="constant.MAGIC_NUMBER.html" title="constant spirv::MAGIC_NUMBER">MAGIC_NUMBER</a></div></li><li><div class="item-name"><a class="constant" href="constant.MAJOR_VERSION.html" title="constant spirv::MAJOR_VERSION">MAJOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.MINOR_VERSION.html" title="constant spirv::MINOR_VERSION">MINOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.REVISION.html" title="constant spirv::REVISION">REVISION</a></div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.Word.html" title="type spirv::Word">Word</a></div></li></ul></section></div></main></body></html>