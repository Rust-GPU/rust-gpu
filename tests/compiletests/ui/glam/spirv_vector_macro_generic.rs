// build-pass
// only-vulkan1.2
// compile-flags: -C target-feature=+GroupNonUniform,+GroupNonUniformShuffleRelative,+ext:SPV_KHR_vulkan_memory_model
// compile-flags: -C llvm-args=--disassemble
// normalize-stderr-test "OpSource .*\n" -> ""
// normalize-stderr-test "OpLine .*\n" -> ""
// normalize-stderr-test "%\d+ = OpString .*\n" -> ""

use spirv_std::Scalar;
use spirv_std::arch::subgroup_shuffle_up;
use spirv_std::glam::Vec3;
use spirv_std::spirv;

#[spirv_std::spirv_vector]
#[derive(Copy, Clone, Default)]
pub struct Vec<T: Scalar> {
    pub x: T,
    pub y: T,
    pub z: T,
}

#[spirv(compute(threads(32)))]
pub fn main(
    #[spirv(storage_buffer, descriptor_set = 0, binding = 0)] input: &Vec<i32>,
    #[spirv(storage_buffer, descriptor_set = 0, binding = 1)] output: &mut Vec<f32>,
) {
    let vec = Vec {
        x: input.x as f32,
        y: input.y as f32,
        z: input.z as f32,
    };
    // any function that accepts a `VectorOrScalar` would do
    *output = subgroup_shuffle_up(vec, 5);
}
