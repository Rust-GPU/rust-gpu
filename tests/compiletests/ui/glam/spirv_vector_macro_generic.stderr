; SPIR-V
; Version: 1.5
; Generator: rspirv
; Bound: 39
OpCapability Shader
OpCapability GroupNonUniform
OpCapability GroupNonUniformShuffleRelative
OpCapability VulkanMemoryModel
OpExtension "SPV_KHR_vulkan_memory_model"
OpMemoryModel Logical Vulkan
OpEntryPoint GLCompute %1 "main" %2 %3
OpExecutionMode %1 LocalSize 32 1 1
OpName %2 "input"
OpName %3 "output"
OpDecorate %6 Block
OpMemberDecorate %6 0 Offset 0
OpDecorate %7 Block
OpMemberDecorate %7 0 Offset 0
OpDecorate %2 NonWritable
OpDecorate %2 Binding 0
OpDecorate %2 DescriptorSet 0
OpDecorate %3 Binding 1
OpDecorate %3 DescriptorSet 0
%8 = OpTypeInt 32 1
%9 = OpTypeVector %8 3
%6 = OpTypeStruct %9
%10 = OpTypePointer StorageBuffer %6
%11 = OpTypeFloat 32
%12 = OpTypeVector %11 3
%7 = OpTypeStruct %12
%13 = OpTypePointer StorageBuffer %7
%14 = OpTypeVoid
%15 = OpTypeFunction %14
%16 = OpTypePointer StorageBuffer %9
%2 = OpVariable  %10  StorageBuffer
%17 = OpTypeInt 32 0
%18 = OpConstant  %17  0
%19 = OpTypePointer StorageBuffer %12
%3 = OpVariable  %13  StorageBuffer
%20 = OpTypePointer StorageBuffer %8
%21 = OpConstant  %17  1
%22 = OpConstant  %17  2
%23 = OpConstant  %17  3
%24 = OpConstant  %17  5
%1 = OpFunction  %14  None %15
%25 = OpLabel
%26 = OpInBoundsAccessChain  %16  %2 %18
%27 = OpInBoundsAccessChain  %19  %3 %18
%28 = OpInBoundsAccessChain  %20  %26 %18
%29 = OpLoad  %8  %28
%30 = OpConvertSToF  %11  %29
%31 = OpInBoundsAccessChain  %20  %26 %21
%32 = OpLoad  %8  %31
%33 = OpConvertSToF  %11  %32
%34 = OpInBoundsAccessChain  %20  %26 %22
%35 = OpLoad  %8  %34
%36 = OpConvertSToF  %11  %35
%37 = OpCompositeConstruct  %12  %30 %33 %36
%38 = OpGroupNonUniformShuffleUp  %12  %23 %37 %24
OpStore %27 %38
OpNoLine
OpReturn
OpFunctionEnd
